File: datasets/C++/Deepseek/DataStructures/hard4.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

struct BitNode {
    unordered_map<int, BitNode*> children;
};

int max_xor_two_numbers(vector<int>& arr) {
    int max_xor = 0;
    BitNode* root = new BitNode();
    
    for (int num : arr) {
        BitNode* node = root;
        BitNode* xor_node = root;
        int current_xor = 0;
        
        for (int i = 31; i >= 0; --i) {
            int bit = (num >> i) & 1;
            // Insert the number into the trie
            if (node->children.find(bit) == node->children.end()) {
                node->children[bit] = new BitNode();
            }
            node = node->children[bit];
            
            // Compute current_xor by choosing the opposite bit if available
            int toggled_bit = 1 - bit;
            if (xor_node->children.find(toggled_bit) != xor_node->children.end()) {
                current_xor += (1 << i);
                xor_node = xor_node->children[toggled_bit];
            } else {
                xor_node = xor_node->children[bit];
            }
        }
        
        if (current_xor > max_xor) {
            max_xor = current_xor;
        }
    }
    
    return max_xor;
}

int main() {
    int n;
    cout << "Enter the number of elements: ";
    cin >> n;
    
    vector<int> arr(n);
    cout << "Enter the elements separated by space: ";
    for (int i = 0; i < n; ++i) {
        cin >> arr[i];
    }
    
    cout << "Maximum XOR: " << max_xor_two_numbers(arr) << endl;
    
    return 0;
}

File: datasets/Python/Deepseek/DataStructures/hard4.py
def max_xor_two_numbers(arr):
    max_xor = 0
    root = {}
    
    for num in arr:
        node = root
        xor_node = root
        current_xor = 0
        
        for i in range(31, -1, -1):
            bit = (num >> i) & 1
            # Insert the number into the trie
            if bit not in node:
                node[bit] = {}
            node = node[bit]
            
            # Compute current_xor by choosing the opposite bit if available
            toggled_bit = 1 - bit
            if toggled_bit in xor_node:
                current_xor += (1 << i)
                xor_node = xor_node[toggled_bit]
            else:
                xor_node = xor_node[bit]
        
        if current_xor > max_xor:
            max_xor = current_xor
    
    return max_xor

def main():
    n = int(input("Enter the number of elements: "))
    arr = list(map(int, input("Enter the elements separated by space: ").split()))
    print("Maximum XOR:", max_xor_two_numbers(arr))

if __name__ == "__main__":
    main()